Source/Core/Common/RangeSizeSet.h

```diff
@@ -0,0 +1,458 @@
+
+    if (from == get_from(bound))
+    {
+      // similarly, if the previous range starts with the given one, just erase until 'to'
+      erase_from_iterator_to_value(bound, to);
+      return;
+    }
+
+    // if we come here, the given range may or may not overlap part of the current 'bound'
+    // (but never the full range), which means we may need to update the end position of it,
+    // or possibly even split it into two
+    if (from < get_to(bound))
+    {
+      if (to < get_to(bound))
+      {
+        // need to split in two
+        bisect_range(bound, from, to);
+      }
+      else
+      {
+        // just update end
+        reduce_to(bound, from);
+      }
+    }
+
+    // and then just erase until to
+    erase_from_iterator_to_value(abound, to);
+    return;
+  }
+
+  void clear()
+  {
+    m_map.clear();
+    m_sizes.clear();
+  }
+
+  size_t size() const { return m_map.size(); }
+
+  const_iterator begin() const { return const_iterator(m_map.begin()); }
+
+  const_iterator end() const { return const_iterator(m_map.end()); }
+
+  by_size_const_iterator by_size_begin() const { return by_size_const_iterator(m_sizes.begin()); }
+
+  by_size_const_iterator by_size_end() const { return by_size_const_iterator(m_sizes.end()); }
+
+  bool operator==(const RangeSizeSet<T>& other) const { return this->m_map == other.m_map; }
+
+  bool operator!=(const RangeSizeSet<T>& other) const { return !(*this == other); }
+
+private:
+  UT calc_size(T from, T to)
+  {
+    // TODO: This should behave correctly when T is signed.
+    return UT(to - from);
+  }
+
+  // Value type stored in the regular range map.
+  struct Value
+  {
+    // End point of the range.
+    T m_to;
+
+    // Pointer to the same range in the by-size multimap.
+    typename std::multimap<UT, typename std::map<T, Value>::iterator, std::greater<UT>>::iterator
+        m_size_it;
+
+    Value(T to) : m_to(to) {}
+
+    bool operator==(const Value& other) const { return this->m_to == other.m_to; }
+
+    bool operator!=(const Value& other) const { return !(*this == other); }
+  };
+
+  // Assumptions that can be made about the data:
+  // - Range are stored in the form [from, to[
+  //   That is, the starting value is inclusive, and the end value is exclusive.
+  // - 'from' is the map key, 'to' is the map value
+  // - 'from' is always smaller than 'to'
+  // - Stored ranges never touch.
+  // - Stored ranges never overlap.
+  std::map<T, Value> m_map;
+
+  // The by-size multimap.
+  // Key is the size of the range.
+  // Value is a pointer to the range in the regular range map.
+  // We use std::greater so that m_sizes.begin() gives us the largest range.
+  std::multimap<UT, typename std::map<T, Value>::iterator, std::greater<UT>> m_sizes;
+
+  T get_from(typename std::map<T, Value>::iterator it) { return it->first; }
+
+  T get_to(typename std::map<T, Value>::iterator it) { return it->second.m_to; }
+
+  typename std::map<T, Value>::iterator insert_range(T from, T to)
```

Comment:
May want to make a local using alias in the private section of the class to prevent needing to continually repeat this typename, etc.